### PR TITLE
Bug 495 card should be added to all tagged decks not just the first one

### LINKED ISSUE #495: [BUG] Obsidian SRS ignores more than one tag
**Describe the bug**
Obsidian SRS ignores more than one tag. It seems as if the algorithm stops assigning a card/note to decks after it matched the first tag it knows.

**To Reproduce**
1. Create this flashcard:

```
---
tags:
  - flashcard
  - flashcard/test
  - flashcard/test2
  - otherFlashcardTag
---
**Spaced repetition** is an **evidence-based** learning technique that is usually performed with **flashcards**.

Newly introduced and more difficult flashcards are shown **more frequently**.

```

2. Add #`otherFlashcardTag` to the list of flashcard tags.
3. Open the review modal.
4. The card will be filed under `flashcard` only.

**Expected behavior**
The card should be IMHO filed under 
  - `flashcard`
  - `flashcard/test`
  - `flashcard/test2` and
  - `otherFlashcardTag`

I would at least expect the card to filed under `flashcard` and `otherFlashcardTag`. It would be best if it belonged to all three decks.

Use case: A flashcard for marketing that is about market research and customer behaviour.
I'd file the card under `#marketing/marketResearch` and `#marketing/customerBehaviour` and maybe `cognitiveBias`. So, I'd like to study the card if I study the `marketing/marketResearch` or the `marketing/customerBehaviour` deck or the whole `marketing` deck if got a lot of time. Furthermore I'd like to see the card if I study `cognitiveBias`es.

**Screenshots**
image
Shows the file and that it only belongs to the `flashcard` deck.
image
List of tags

**Versions (please complete the following information):**

-   OS: `Manjaro Linux`
-   Obsidian version: `v0.15.9`
-   Plugin version: `v1.8.0`
-   If on desktop, Installer version: `5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) obsidian/0.15.9 Chrome/100.0.4896.160 Electron/18.3.5 Safari/537.36`

### PR DESCRIPTION
**Pull request by MostlyArmless**
I have a change which fixes #495, allowing a card to be added simultaneously to multiple decks if the note is tagged with multiple valid deck tags.

This does lead to some questionable behavior, where if a user tags the same note with `#parentTag` and `#parentTag/childTag`, then the cards in that note will be single-counted in the `#parentTag/childTag` deck (which is fine), but double-counted in the `#parentTag` deck (which seems weird, but hey it's what they've literally asked for by doing that).

# Example of what I mean

Here's the contents of the note:
image

and here's the resulting UI from this plugin when you use the 'cram flashcards in this note' command:
image

I'm not sure if this is how we want the plugin to behave, or if we would perhaps like to make it configurable by having a setting that determines what happens in this situation. I'm open to suggestions, as I haven't been using the `obsidian-spaced-repetition` plugin personally for very long so I don't have a good feel for what makes sense here.

### CODE REVIEW

File: src/main.ts [hunk 1]
```diff
@@ -586,34 +586,33 @@ export default class SRPlugin extends Plugin {
...
+            const tagsInThisNote = getAllTags(fileCachedData) || [];
 
-            outer: for (const tagToReview of this.data.settings.flashcardTags) {
-                for (const tag of tags) {
+            for (const tagToReview of this.data.settings.flashcardTags) {
+                for (const tag of tagsInThisNote) {
                     if (tag === tagToReview || tag.startsWith(tagToReview + "/")) {
-                        deckPath = tag.substring(1).split("/");
-                        break outer;
+                        deckPaths.push(tag.substring(1).split("/"));
```
Comment (MostlyArmless) [inline src/main.ts:604]: This is the crux of the change. Just append to an array here instead of early-returning from the loop.

**Comment by st3v3nmw** [comment]
> This does lead to some questionable behavior, where if a user tags the same note with `#parentTag` and `#parentTag/childTag`, then the cards in that note will be single-counted in the `#parentTag/childTag` deck (which is fine), but double-counted in the `#parentTag` deck (which seems weird, but hey it's what they've literally asked for by doing that).

Yeah, this approach would break a few things like the counting as you've mentioned and we'd have the card in multiple decks, so if we review it the first time in deck A, the same card will show up later in the same review session (in deck A/B).

We'll probably need to rewrite parts of this plugin to use hash(cardText) as IDs to make it easier to remove/pick/update the same card in multiple decks

**Review by st3v3nmw — COMMENTED** [review]
Hi, thanks for the contribution 😄!

Sorry the underlying code has gone through a major refactor recently. I'm not sure if you're still available to help resolve the merge conflicts.

Stephen,

**Comment by pikatwinky** [comment]
Any chances to check this merge conflicts @MostlyArmless ? This PR looks really really promising...

**Comment by MostlyArmless** [comment]
Sure I'll take a look tonight.

**Comment by ronzulu** [comment]
Hi @4Source 

Yes, if actually been quite involved, but I'm at the tail end. I'm hoping to release a beta version in the next couple of days. Will you be able to test and give me feedback?

**Comment by 4Source** [comment]
I think that should be possible. I'll wait for the release. Will you mark this here when the time comes?

**Comment by ronzulu** [comment]
Hi @4Source @MostlyArmless @pikatwinky (and anyone else)

I have completed the update but would appreciate beta testing and feedback before I finalise the PR.

The draft PR together with the updated code is available at:
https://github.com/st3v3nmw/obsidian-spaced-repetition/pull/834

Cheers
Ronny

**Comment by ronzulu** [comment]
Hi @MostlyArmless have you had a chance to look at https://github.com/st3v3nmw/obsidian-spaced-repetition/pull/834?

After all your good work with this PR, I want to make sure you are happy with the final result.

Cheers
Ronny

**Comment by ronzulu** [comment]
This has been superseded by #834, which was based on the refactored code base.

**Pull request closed without merging**